### DISCUSSION TITLE
When i try to use my custom (client only) commands in 1.19, it keeps saying that it is an incompleete/invalid command.

### DISCUSSION THREAD

**Discussion by AgentKosticka** [FabricMC/fabric-api#2335] [Mod Dev Support] [answered]
I have this code right here: 
`public static void RegisterCommands() {
        ClientCommandRegistrationCallback.EVENT.register((dispatcher, registryAccess) ->
                dispatcher.register(literal("something")
                        .then(literal("somethingelse")
                            .executes(MyFile.SomeFunctionHere()))));
    }`
can someone help me, what is wrong with this code why it keeps saying that. If you coment anything it would mean a HUGE help to me. Thanks

**Comment by warjort** [comment, marked as answer]
Let me format it like this
````
dispatcher.register(
    literal("something") .then(
        literal("somethingelse") 
        .executes(MyFile.SomeFunctionHere())
    )
)
````

So you can see you have a command that needs 2 literals. You would enter it as
````/something somethingelse````

You know you can use tab to autocomplete commands or give suggested completions?

**Reply by AgentKosticka** [reply]
here's in game footage of the problem:
[2022-06-18 12-49-29.zip](https://github.com/FabricMC/fabric/files/8933401/2022-06-18.12-49-29.zip)
 (vid is 10.9 megabytes so i packed it)

**Reply by warjort** [reply]
What does your method return? 
I can get the same behaviour if I do
````
 .executes(null)
````
i.e. I say there is no command

Try changing it to use this simple lambda that should print to the terminal
````
 .executes(context -> {
    System.out.println("Running command");
    return 0;
 })
````

**Reply by AgentKosticka** [reply]
Method writes something in the logger and returns null. 
So make the method not void and not return null? (ex: int, 0)

**Reply by AgentKosticka** [reply]
Btw what does `context - > blablabla` mean?

**Reply by warjort** [reply]
The code you have written says I am going to use whatever SomeFunctionHere returns as the command to execute.
You say you are returning null which tells minecraft there is no such command.
````
 .executes(MyFile.SomeFunctionHere()) // means call the method then use the result as the command
````
If you actually want SomeFunctionHere as the command then you would write
````
 .executes(MyFile::SomeFunctionHere) // means use that function as the command to execute
````
this is a method reference not a method call like you have.
But you will need to make sure SomeFunctionHere has the correct method signature for a command.

> Btw what does context - > blablabla mean?

Its an anonymous function - more commonly known as a lambda function.
In this case a function that takes one parameter - the context
https://docs.oracle.com/javase/tutorial/java/javaOO/lambdaexpressions.html
So instead of giving it the method reference to elsewhere, the function is defined inline.

**Reply by Sjouwer** [reply]
I recommend this guide which explains how commands work:
https://fabricmc.net/wiki/tutorial:commands

**Reply by AgentKosticka** [reply]
> 

Here is another footage of me trying to use `(MyFile::SomeFunctionHere)`:


https://user-images.githubusercontent.com/79516204/174451139-50797593-8a4c-46e0-9c98-e6fa3c49661e.mp4

Still can't figure out whats wrong... :(

**Reply by warjort** [reply]
Like I said above, if you want to use method references:
````But you will need to make sure SomeFunctionHere has the correct method signature for a command.````
So your method's signature for a client command needs to look like this:
````
	public static int something(CommandContext<FabricClientCommandSource> context) {
		LOGGER.info("Something");
		return 0;
	}
````
This is the named version of the anonymous lambda I posted before.

**Comment by warjort** [comment]
Aside:
In the latest video you posted, you show the client commands getting registered from your mod initializer and you are placing a client command method in your normal mod code.
This is incorrect. The client command stuff is only available on the client. You need to put it in a client mod initializer
https://fabricmc.net/wiki/documentation:entrypoint

If you leave it as it is, it will crash if your try to load your mod on a dedicated server.

**Reply by AgentKosticka** [reply]
Btw not related to the problem, but I am trying to implement a way to send text message in the chat visible only to the client. I found this example: https://github.com/FabLabsMC/ClientCommands/blob/162432f59c7653380c4f1c78e755855d1ec8b54c/src/main/java/io/github/fablabsmc/fablabs/mixin/command/ClientCommandSourceMixin.java
But it keeps saying that it douesnt know, what `client` is, or what `.addChatMessage` is

**Reply by warjort** [reply]
You should use the source of the context passed to your command.
https://github.com/FabricMC/fabric/blob/1.19/fabric-command-api-v2/src/client/java/net/fabricmc/fabric/api/client/command/v2/FabricClientCommandSource.java

Think of it as the player, but be aware for server commands it could be a command block.

So you would do something like:
````
	public static int something(CommandContext<FabricClientCommandSource> context) {
		context.getSource().sendFeedback(Text.literal("Hello"));
		return 0;
	}
````

**Reply by AgentKosticka** [reply]
this method worked but somehow added 2 new commands (/fcc and /fabric-context-blablabal(dont remember)) Do you know a fix?

**Reply by warjort** [reply]
Thats just a helper method so you can list just the client only commands by typing ````/fcc help```` or the longer name.
https://github.com/FabricMC/fabric/blob/1c5dd50b652b123aaa0ed10c58dcd5c15e58b855/fabric-command-api-v2/src/client/java/net/fabricmc/fabric/impl/command/client/ClientCommandInternals.java#L141

**Reply by AgentKosticka** [reply]
ok and how do i get rid of it

> Thats just a helper method so you can list just the client only commands by typing `/fcc help` or the longer name.
> 
> https://github.com/FabricMC/fabric/blob/1c5dd50b652b123aaa0ed10c58dcd5c15e58b855/fabric-command-api-v2/src/client/java/net/fabricmc/fabric/impl/command/client/ClientCommandInternals.java#L141
EDIT: I dont understand java very well, but other client command mods do not include /fcc so there must be a way to get rid of it

**Reply by warjort** [reply]
Its not your command. Why do you want to get rid of it?
Its just the same as typing ````/help```` to list all commands.

I'm sure there are ways to remove other people's commands but I doubt it is straightforward, its not a normal operation. 
I'll leave that for you to research, it will involve getting hold of the root command node from the command dispatcher and removing the data from it, maybe it will need a mixin to make it mutable?

**Reply by warjort** [reply]
> I dont understand java very well, but other client command mods do not include /fcc so there must be a way to get rid of it

Its added by fabric "automagically" when anybody registers a client command. So yes it will be there if other mods have client commands.

**Reply by AgentKosticka** [reply]
ok, thanks for the help. But you understand my situation, i want a peaceful mod and some random UGLY command (/fabric-command-api-v2:client) just apears in my mod out of thin air

**Reply by Juuxel** [reply]
It's not a "random UGLY command", it's the client command version of `/help` and cannot be disabled, just like you can't disable the serverside help command.

The reason why it's not always there to begin with is that it's only needed if client commands are registered.
